### PR TITLE
Support gRPC TLS communication and TLS registry and FW support for gRPC on same server

### PR DESCRIPTION
**Pull request by michalvavrik**
### Summary

- Adds support for gRPC running on same Vert.X HTTP server (which shall be primary Quarkus way in the future, not legacy separate server).
- Closes: #1048
- Closes: #1054
- Adds support for TLS and mTLS communication with gRPC service
- Integrates TLS registry support
- I needed to use Netty directly (which according to my debugging was used anyway, it is one of hardcoded classes in gRPC core and it was selected by default provider in cases I debugged without changes in this PR). This allows me to work with SSL context.
- I had issue with managed context, which I can see in our TS we also had, see https://github.com/quarkus-qe/quarkus-test-suite/blob/38cdbd539b172db0d9e6334759df3487ac8864fe/http/grpc/src/test/java/io/quarkus/ts/http/grpc/XeparateServerIT.java#L13 so from now on, I am returning managed channel has method scope. I looked how is it done upstream and I can see using `AfterEach` and `BeforeEach` like here: https://github.com/quarkusio/quarkus/blob/main/integration-tests/grpc-mutual-auth/src/test/java/io/quarkus/grpc/examples/hello/HelloWorldMutualTlsServiceTest.java
- in regards of Quarkus QE Test Suite `http/grpc` module, I prepared changes in TS here https://github.com/quarkus-qe/quarkus-test-suite/commit/14fdffba63042b165a7e5dc5fd571f8a9f17387f (that commit also adapts TS to other breaking changes I introduced in my other PRs before this one) and tested it with them:
  - seems to fix issues with closing channel, so that we can rename `XeparateServer` to `SeparateServer` (I only run it few times, I don't swear on it...)
  - `SameServerIT` https://github.com/quarkus-qe/quarkus-test-suite/blob/main/http/grpc/src/test/java/io/quarkus/ts/http/grpc/SameServerIT.java had to work around FW support for same server, this fixes it

Please check the relevant options

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Release (follows conventions described in the [RELEASE.md](https://github.com/quarkus-qe/quarkus-test-framework/blob/main/RELEASE.md))
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update
- [ ] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Example scenarios has been updated / added
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

**Comment by michalvavrik** [comment]
Issues on Windows are related to escaping, I don't know if I fixed them as obviously I didn't try it (it's bit late...). If it's red tomorrow, I'll work around that, but it's a small thing, I think you can freely comment on changes. Thanks

**Comment by michalvavrik** [comment]
Interesting, Quarkus turns `-Dquarkus.tls.grpc-tls.trust-store.pem.certs="C:\Users\RUNNER~1\AppData\Local\Temp\quarkus-qe-certs809756706346777605\quarkus-qe-server-ca.crt"` into:

```
2024-09-02T21:43:49.8846143Z 21:43:49,799 INFO  [app] Caused by: java.nio.file.NoSuchFileException: C:UsersRUNNER~1AppDataLocalTempquarkus-qe-certs809756706346777605quarkus-qe-server-ca.crt
2024-09-02T21:43:49.8847380Z 21:43:49,799 INFO  [app] 	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
2024-09-02T21:43:49.8848370Z 21:43:49,799 INFO  [app] 	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
2024-09-02T21:43:49.8849385Z 21:43:49,799 INFO  [app] 	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
2024-09-02T21:43:49.8850459Z 21:43:49,799 INFO  [app] 	at java.base/sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:236)
2024-09-02T21:43:49.8851372Z 21:43:49,799 INFO  [app] 	at java.base/java.nio.file.Files.newByteChannel(Files.java:380)
2024-09-02T21:43:49.8852109Z 21:43:49,799 INFO  [app] 	at java.base/java.nio.file.Files.newByteChannel(Files.java:432)
2024-09-02T21:43:49.8852987Z 21:43:49,799 INFO  [app] 	at java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:422)
2024-09-02T21:43:49.8853880Z 21:43:49,799 INFO  [app] 	at java.base/java.nio.file.Files.newInputStream(Files.java:160)
2024-09-02T21:43:49.8854673Z 21:43:49,799 INFO  [app] 	at io.quarkus.tls.runtime.config.TlsConfigUtils.read(TlsConfigUtils.java:43)
2024-09-02T21:43:49.8855286Z 21:43:49,799 INFO  [app] 	... 20 more
```
I'll play with it tomorrow.

**Comment by michalvavrik** [comment]
Starting experiments on Windows, will report back when I fix it.

**Comment by michalvavrik** [comment]
I have doubled backslashes for my use case, which fixed the issue. I think we need to do it for everywhere https://github.com/quarkus-qe/quarkus-test-framework/issues/1275 but I do not there to do this in here. I am in hurry and don't have time to fix/debug side-effects in backports etc.

File: quarkus-test-service-grpc/src/main/java/io/quarkus/test/bootstrap/GrpcService.java [hunk 1]
```diff
@@ -1,16 +1,100 @@
...
+        return new CloseableManagedChannel(nettyChannel);
+    }
+
+    public CloseableManagedChannel securedGrpcChannel() {
+        return securedGrpcChannel(null);
+    }
+
+    public CloseableManagedChannel securedGrpcChannel(String clientCnName) {
+        if (QuarkusProperties.useSeparateGrpcServer(context)) {
+            throw new IllegalStateException("TLS support for separate gRPC server is currently not supported");
```
Comment (fedinskiy) [inline quarkus-test-service-grpc/src/main/java/io/quarkus/test/bootstrap/GrpcService.java:32]:
```suggestion
            throw new IllegalStateException("TLS is not currently supported for a separate gRPC server");
```

File: quarkus-test-service-grpc/src/main/java/io/quarkus/test/bootstrap/GrpcService.java [hunk 2]
```diff
@@ -1,16 +1,100 @@
...
+    }
+
+    private SslContext createSslContext(String clientCnName) {
+        SslContextBuilder sslContextBuilder = GrpcSslContexts.forClient();
+
+        // only implementing this for one certificate and one client certificate for we don't need more ATM
+        CertificateBuilder certBuilder = getPropertyFromContext(CertificateBuilder.INSTANCE_KEY);
+
+        if (certBuilder.certificates().size() != 1) {
+            throw new IllegalStateException("Exactly one certificate must exist for the SSL configuration to work");
```
Comment (fedinskiy) [inline quarkus-test-service-grpc/src/main/java/io/quarkus/test/bootstrap/GrpcService.java:49]: We should mention, which property we used to look for a certificate  in the error message. And a number of found certificates

File: examples/grpc/src/test/java/io/quarkus/qe/grpc/GrpcMtlsTlsRegistryIT.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import io.quarkus.test.services.Certificate.ClientCertificate;
+import io.quarkus.test.services.QuarkusApplication;
+
+@QuarkusScenario
+public class GrpcMtlsTlsRegistryIT {
+
+    private static final String CN = "Hagrid";
+    private static final String NAME = "Albus";
+
+    @QuarkusApplication(grpc = true, ssl = true, certificates = @Certificate(format = PEM, configureHttpServer = true, configureKeystore = true, configureTruststore = true, tlsConfigName = "grpc-tls", clientCertificates = @ClientCertificate(cnAttribute = CN)))
```
Comment (fedinskiy) [inline examples/grpc/src/test/java/io/quarkus/qe/grpc/GrpcMtlsTlsRegistryIT.java:20]: As far as I understand, `CN=Hagrid` is now included in certificates. Please, add the scripts you used to generate them

**Review by fedinskiy — CHANGES_REQUESTED** [review]
We need a couple of small changes.

And what about Openshift? Will it work there?

**Comment by michalvavrik** [comment]
> And what about Openshift? Will it work there?

https://github.com/quarkus-qe/quarkus-test-framework/issues/1053 IIRC you cannot have both enabled HTTP2 and HTTP1 in OpenShift cluster. Things might have changed, but it's beyond what I can do ATM.

File: quarkus-test-service-grpc/src/main/java/io/quarkus/test/bootstrap/GrpcService.java (hunk 2 above)
Comment (michalvavrik) [reply quarkus-test-service-grpc/src/main/java/io/quarkus/test/bootstrap/GrpcService.java:49]: will do

File: examples/grpc/src/test/java/io/quarkus/qe/grpc/GrpcMtlsTlsRegistryIT.java (hunk 3 above)
Comment (michalvavrik) [reply examples/grpc/src/test/java/io/quarkus/qe/grpc/GrpcMtlsTlsRegistryIT.java:20]: > As far as I understand, CN=Hagrid is now included in certificates. Please, add the scripts you used to generate them

yeah, you can see that `SecurityIdentity` is not anonymous and its principal contains DN with `CN=Hagrid` which is then asserted in `HiReply`. therefore authentication is verified and so is certificates value

**Review by fedinskiy — APPROVED** [review]

**Pull request merged**